src/rewrite.spec.ts

```diff
@@ -0,0 +1,80 @@
+import { RewriteRule, RewriteRules } from "./rewrite";
+
+describe("RewriteRule", () => {
+  describe("without capture groups", () => {
+    it("applies the expected changes", () => {
+      const rule = new RewriteRule(/-[a-z0-9]+-?/gi, "");
+      expect(rule.apply("")).toEqual("");
+      expect(rule.apply("chicken")).toEqual("chicken");
+      expect(rule.apply("one-TWO-thRee-fOUr")).toEqual("onethRee");
+    });
+  });
+
+  describe("with capture groups", () => {
+    it("applies the expected changes", () => {
+      const rule1 = new RewriteRule(/\$([0-9]+)(\.[0-9]+)?/g, "£$1");
+      expect(rule1.apply("")).toEqual("");
+      expect(rule1.apply("chicken")).toEqual("chicken");
+      expect(rule1.apply("They are $5, $7.90, and $1024.9876.")).toEqual(
+        "They are £5, £7, and £1024."
+      );
+
+      const rule2 = new RewriteRule(
+        /-[0-9a-f]{8}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{12}(@example.com)$/gi,
+        "$1"
+      );
+      expect(
+        rule2.apply(
+          "jane.doe-some-test-6f82fbbe-d36a-4c5c-b47b-84100122fbbc@example.com"
+        )
+      ).toEqual("jane.doe-some-test@example.com");
+      expect(
+        rule2.apply(
+          "jane.doe-some-test-6F82FBBE-D36A-4C5C-B47B-84100122FBBC@example.com"
+        )
+      ).toEqual("jane.doe-some-test@example.com");
+    });
+  });
+});
+
+describe("RewriteRules", () => {
+  describe("when there are no rules", () => {
+    const rules = new RewriteRules();
+
+    it("does not mutate the given value at all", () => {
+      expect(rules.apply("chicken")).toEqual("chicken");
+    });
+  });
+
+  describe("when there is a simple rewrite rule", () => {
+    const rules = new RewriteRules().appendRule(new RewriteRule(/dog/, "cat"));
+
+    it("applies the expected changes on basic data types", () => {
+      expect(rules.apply("The dog dodges the doggie.")).toEqual(
+        "The cat dodges the doggie."
```

Comment:
**Question:** Is this output expected? I would have thought this would output `The cat dodges the catgie`

In the below test for an object we replace `dog` within a string (e.g. `dogs` becomes `cats`)
![image](https://github.com/airtasker/proxay/assets/22439259/431dd8b8-84ed-44a8-90cf-5cc2f986d164)

Reply:
Ah I see - it doesn't have the global replace symbol. Looks good 👍